### ISSUE TITLE
Boolean operators not working for dynamic objects

### ISSUE THREAD

**Issue by AndiVarga** [davideicardi/DynamicExpresso#162]
Hello,

I've encountered this issue when trying to use boolean operators with dynamic properties:

```
dynamic dyn = new ExpandoObject();
dyn.val1 = 100;
dyn.val2 = 200;

var interpreter = new Interpreter();
interpreter.SetVariable("dyn", dyn);

interpreter.Eval("dyn.val1 == 100"); //returns true, as expected
interpreter.Eval("dyn.val1 == 100 && dyn.val2 == 200"); // throws exception: "Invalid argument value\r\nParameter name: operation
```

Please advise if this is indeed a bug or something is not properly configured.

**Comment by davideicardi** [comment]
Hi @AndiVarga . Yes it looks like a bug.

You are working with the latest version?

Can you try to reproduce it in standard C# code? I would expect that it should compile, but just to be sure ...

**Comment by davideicardi** [comment]
Yes, it looks like a bug.
As usual any help/PR is appreciated ;-) .

**Comment by rotsin** [comment]
In 

```
private Expression ParseConditionalAnd()
{
	var left = ParseLogicalOr();
	while (_token.id == TokenId.DoubleAmphersand)
	{
		NextToken();
		var right = ParseLogicalOr();
		CheckAndPromoteOperands(typeof(ParseSignatures.ILogicalSignatures), ref left, ref right);
		left = GenerateBinary(ExpressionType.AndAlso, left, right);
	}
	return left;
}
```

which is called when there are dynamic expressions, expression type AndAlso is passed, problem is that Microsoft.CSharp.RuntimeBinder.Binder.BinaryOperation does not seem to support this type of operation. Changing it to And seems to resolve the issue, also added a small test to check that Andi's scenario works and it does, and all other tests run successfully.

Cound not push my branch to the repo (I may need rights), but the fix should be reasonably simple that there is no need to get my on the PR pipeline.

**Comment by metoule** [comment]
Changing `AndAlso` to `And` would impact all operations that rely on the "short-circuiting" logical AND operator, which doesn't seem like a good idea. It would mean that all `&&` operations are considered as `&`, which will lead to unexpected side effects.

For example:

```` 
string test = null;
var res = test != null && test.IsNormalized();
````

using `AndAlso` ensures that the second operand is not evaluated, and therefore there's no `NullReferenceException`. Using `And` would lead to an unexcepted exception.

**Comment by rotsin** [comment]
Hey @metoule, good point! It would only have impacted the dynamics path, which for those scenarios it is not working currently. At any rate, I have looked into the source code of the runtime binder and indeed it seems that AndAlso and OrElse are explicitly not supported for it.

This means that for dynamics, the binder is a no go. A simple but hackish solution could be something like:

`		private static Expression GenerateBinaryDynamic(ExpressionType binaryType,Expression left, Expression right)
		{
			switch (binaryType)
			{
				case ExpressionType.AndAlso:
					return Expression.AndAlso(Expression.Convert(left, typeof(bool)), Expression.Convert(right, typeof(bool)));
				case ExpressionType.OrElse:
					return Expression.OrElse(Expression.Convert(left, typeof(bool)), Expression.Convert(right, typeof(bool)));
				default:
					{
						var binder = Microsoft.CSharp.RuntimeBinder.Binder.BinaryOperation(
										CSharpBinderFlags.None,
										binaryType,
										typeof(Parser),
										new CSharpArgumentInfo[] {
												CSharpArgumentInfo.Create(CSharpArgumentInfoFlags.None, null),
												CSharpArgumentInfo.Create(CSharpArgumentInfoFlags.None, null)}
										);

						return Expression.Dynamic(binder, typeof(object), left, right);
					}
			}
		}`

Since these two scenarios are not supported in runtime binder, treat them separately. The only issue is, do we expect in some scenarios the left or right in an && or || to not be convertible to bool?

**Comment by davideicardi** [comment]
Closed by #163 .

Thank you @rotsin , just merged PR.
